File: solutions/3375-kth-smallest-amount-with-single-denomination-combination/kth-smallest-amount-with-single-denomination-combination.c
long long gcd(long long a, long long b) {
    while (b != 0) {
        long long t = b;
        b = a % b;
        a = t;
    }
    return a;
}

int trailingZeros(int x) {
    int count = 0;
    while ((x & 1) == 0) {
        count++;
        x >>= 1;
    }
    return count;
}

int popcount(int x) {
    int count = 0;
    while (x) {
        count += x & 1;
        x >>= 1;
    }
    return count;
}

long long count(long long x, int m, long long* lcm) {
    long long res = 0;
    for (int mask = 1; mask < m; mask++) {
        if (lcm[mask] > x) continue;

        if (popcount(mask) & 1) {
            res += x / lcm[mask];
        } else {
            res -= x / lcm[mask];
        }
    }
    return res;
}

int compare(const void* a, const void* b) { return (*(int*)a - *(int*)b); }

long long findKthSmallest(int* coins, int coinsSize, int k) {
    qsort(coins, coinsSize, sizeof(int), compare);
    int* newCoins = (int*)malloc(coinsSize * sizeof(int));
    int newSize = 0;
    for (int i = 0; i < coinsSize; i++) {
        int flag = 1;
        for (int j = 0; j < newSize; j++) {
            if (coins[i] % newCoins[j] == 0) {
                flag = 0;
                break;
            }
        }
        if (flag) {
            newCoins[newSize++] = coins[i];
        }
    }

    int n = newSize;
    int m = 1 << n;
    long long* lcm = (long long*)malloc(m * sizeof(long long));
    long long l = k;
    long long r = (long long)newCoins[0] * k + 1;

    lcm[0] = 1;
    for (int mask = 1; mask < m; mask++) {
        int preMask = mask & (mask - 1);
        int i = trailingZeros(mask);

        long long tmp = lcm[preMask] / gcd(lcm[preMask], newCoins[i]);
        if (tmp <= r / newCoins[i]) {
            lcm[mask] = tmp * newCoins[i];
        } else {
            lcm[mask] = r + 1;
        }
    }

    while (l < r) {
        long long x = l + (r - l) / 2;
        if (count(x, m, lcm) >= k) {
            r = x;
        } else {
            l = x + 1;
        }
    }

    long long result = l;
    free(newCoins);
    free(lcm);
    return result;
}

File: solutions/3375-kth-smallest-amount-with-single-denomination-combination/kth-smallest-amount-with-single-denomination-combination.ts
function findKthSmallest(coins: number[], k: number): number {
    coins.sort((a, b) => a - b);
    const newCoins: number[] = [];
    for (const x of coins) {
        let flag: boolean = true;
        for (const y of newCoins) {
            if (x % y === 0) {
                flag = false;
                break;
            }
        }
        if (flag) {
            newCoins.push(x);
        }
    }
    coins = newCoins;

    const n: number = coins.length;
    const m: number = 1 << n;
    const lcm: bigint[] = new Array(m).fill(0n);
    let l: bigint = BigInt(k);
    let r: bigint = BigInt(coins[0]) * BigInt(k) + 1n;

    const gcd = (a: bigint, b: bigint): bigint => {
        a = a < 0n ? -a : a;
        b = b < 0n ? -b : b;
        while (b !== 0n) {
            [a, b] = [b, a % b];
        }
        return a;
    };

    const trailingZeros = (x: number): number => {
        let count = 0;
        while ((x & 1) === 0) {
            count++;
            x >>= 1;
        }
        return count;
    };

    const popcount = (x: number): number => {
        let count = 0;
        while (x) {
            count += x & 1;
            x >>= 1;
        }
        return count;
    };

    lcm[0] = 1n;
    for (let mask = 1; mask < m; mask++) {
        const preMask: number = mask & (mask - 1);
        const i: number = trailingZeros(mask);

        const coin: bigint = BigInt(coins[i]);
        const tmp: bigint = lcm[preMask] / gcd(lcm[preMask], coin);
        if (tmp <= r / coin) {
            lcm[mask] = tmp * coin;
        } else {
            lcm[mask] = r + 1n;
        }
    }

    const count = (x: bigint): bigint => {
        let res: bigint = 0n;
        for (let mask = 1; mask < m; mask++) {
            if (lcm[mask] > x) continue;

            if (popcount(mask) & 1) {
                res += x / lcm[mask];
            } else {
                res -= x / lcm[mask];
            }
        }
        return res;
    };

    while (l < r) {
        const mid: bigint = (l + r) / 2n;
        if (count(mid) >= k) {
            r = mid;
        } else {
            l = mid + 1n;
        }
    }
    return Number(l);
}